### ISSUE TITLE
Dozer and Singletons 

### ISSUE THREAD

**Issue by orange-buffalo** [DozerMapper/dozer#375]
Currently Dozer is built on singletons:
* `org.dozer.DozerInitializer#instance`
* `org.dozer.config.BeanContainer#instance`
* `org.dozer.loader.MappingsParser#INSTANCE`
* `org.dozer.config.GlobalSettings#instance`
* `org.dozer.loader.xml.XMLParserFactory#instance`
* `org.dozer.stats.GlobalStatistics#singleton`
* others?

Besides obvious problems like tight coupling between classes and impossibility to replace dependencies in tests, this also causes issues in OSGi environment (see #22).

The need to have global state for Dozer is questionable. Alternative solution would be to remove singletons and inject dependencies via constructors. Then every created instance of `Mapper` could be completely independent of others, and dependencies could be easily mocked for better testing. I believe #34 was created to move into this direction.

If it is agreed that banishing singletons is beneficial, I have a suggestion on customization API. The basic idea is inspired by [Apache HttpClientBuilder](http://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/impl/client/HttpClientBuilder.html). We could create `DozerMapperBuilder`, which provides methods for customizing any Dozer component, handles defaults and creates instance of `Mapper` by constructing, injecting and wiring all the components together. 
Then instead of calling 
```java
    Mapper mapper = new DozerBeanMapper(Arrays.asList(myMappingFile, anotherMappingFile));
```
clients would use the new builder:
```java
    Mapper mapper = DozerMapperBuilder.create()
            .addMappingFile(myMappingFile)
            .addMappingFile(anotherMappingFile)
            .build();
```

We could also replace`DozerBeanMapperSingletonWrapper` with declarative API:
```java
    Mapper mapper = DozerMapperBuilder.buildDefault();
``` 

Examples of customization via Builder API would be:
```java
    Mapper mapper = DozerMapperBuilder.create()
        .addMappingFile(myMappingFile)
        .setProxyResolver(new MyResolver())
        .setClassLoader(mySpecificClassLoader)
        .setElEngine(ShinyElImplementation.class)
        .build()  
```

@garethahealy, what do you think, does it make sense to start working on this idea?

**Comment by garethahealy** [comment]
@orange-buffalo ; moving away from singletons is definitely the aim for the future. More of a styling point from the above, but it would be nice to:

- Make DozerBeanMapper private
- DozerBeanMapper then includes a static builder class, which is used to create the mapper
- Use of fluent style API ++
- Use preposition keywords for fluent API, i.e.: withElEngine instead of setElEngine

i.e.: https://github.com/hawkular/hawkular-metrics/blob/master/core/metrics-model/src/main/java/org/hawkular/metrics/model/NumericBucketPoint.java

Since this is for > 6.1 release - _i would expect anyways_ - i'll raise another issue to track change proposals and link back.

I would also suggest attempting to break this down into several changes if possible, just to make reviewing easier.

**Comment by orange-buffalo** [comment]
@garethahealy Agree with all proposals, the only thing that if the builder is big enough, it might be reasonable to make it a top-level class while making constructor of `DozerBeanMapper` package-private instead of private. But let's see how it goes.

**Comment by garethahealy** [comment]
@orange-buffalo ; sounds good to me.

**Comment by orange-buffalo** [comment]
Closed via #403.
